compiler/src/symboltable.rs

```diff
@@ -193,71 +200,137 @@ impl<'a> SymbolTableAnalyzer<'a> {
         for sub_table in sub_tables {
             self.analyze_symbol_table(sub_table)?;
         }
-        let (symbols, _) = self.tables.pop().unwrap();
+        let (symbols, st_typ) = self.tables.pop().unwrap();
 
         // Analyze symbols:
         for symbol in symbols.values_mut() {
-            self.analyze_symbol(symbol)?;
+            self.analyze_symbol(symbol, st_typ)?;
         }
-
         Ok(())
     }
 
-    fn analyze_symbol(&self, symbol: &mut Symbol) -> SymbolTableResult {
-        match symbol.scope {
-            SymbolScope::Nonlocal => {
-                // check if name is defined in parent table!
-                let parent_symbol_table = self.tables.last();
-                // symbol.table.borrow().parent.clone();
-
-                if let Some((symbols, _)) = parent_symbol_table {
-                    let scope_depth = self.tables.len();
-                    if !symbols.contains_key(&symbol.name) || scope_depth < 2 {
+    fn analyze_symbol(
+        &mut self,
+        symbol: &mut Symbol,
+        curr_st_typ: SymbolTableType,
+    ) -> SymbolTableResult {
+        if symbol.is_assign_namedexpr_in_comprehension
+            && curr_st_typ == SymbolTableType::Comprehension
+        {
+            self.analyze_symbol_comprehension(symbol, 0)?
```

Comment:
You could avoid putting the rest of the function in an indented block by just returning here, and leaving out the else:
```rust
if ... {
  return self.analyze_...()
}
match symbol.scope {...}
```

Reply:
While it would reduce the indentation, it also adds a return statement, and functions with multiple return statements are arguably harder to reason about.

My question is why the field `is_assign_namedexpr_in_comprehension` is required. I do not fully grasp this idea.

Reply:
You can see and argument it in both directions, under pragmatic ocnsideration such as under dogmatic considerations. Usually have the tendency to use early returns in case of some pre checks and otherwise use nesting. I guess this still readable although the nesting level is high. And the most helpful optimization in terms of readablity is to move the whole else branch into a dedicated function.